server/src/main/java/org/elasticsearch/action/admin/indices/dangling/TransportDeleteDanglingIndexAction.java

```diff
@@ -0,0 +1,207 @@
+        );
+        this.settings = settings;
+        this.nodeClient = nodeClient;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected DeleteDanglingIndexResponse read(StreamInput in) throws IOException {
+        return new DeleteDanglingIndexResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        DeleteDanglingIndexRequest deleteRequest,
+        ClusterState state,
+        ActionListener<DeleteDanglingIndexResponse> deleteListener
+    ) throws Exception {
+        findDanglingIndex(deleteRequest.getIndexUUID(), new ActionListener<>() {
+
+            @Override
+            public void onResponse(Index indexToDelete) {
+                // This flag is checked at this point so that we always check that the supplied index ID
+                // does correspond to a dangling index.
+                if (deleteRequest.isAcceptDataLoss() == false) {
+                    deleteListener.onFailure(new IllegalArgumentException("accept_data_loss must be set to true"));
+                    return;
+                }
+
+                String indexName = indexToDelete.getName();
+
+                final ActionListener<DeleteDanglingIndexResponse> clusterStateUpdatedListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(DeleteDanglingIndexResponse response) {
+                        deleteListener.onResponse(response);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.debug("Failed to delete dangling index [{}]" + indexName, e);
+                        deleteListener.onFailure(e);
+                    }
+                };
+
+                clusterService.submitStateUpdateTask(
+                    "delete-dangling-index " + indexName,
+                    new AckedClusterStateUpdateTask<>(deleteRequest, clusterStateUpdatedListener) {
+
+                        @Override
+                        protected DeleteDanglingIndexResponse newResponse(boolean acknowledged) {
+                            return new DeleteDanglingIndexResponse(acknowledged);
+                        }
+
+                        @Override
+                        public ClusterState execute(final ClusterState currentState) {
+                            return deleteDanglingIndex(currentState, indexToDelete);
+                        }
+                    }
+                );
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug("Failed to list dangling indices", e);
+                deleteListener.onFailure(e);
+            }
+        });
+    }
+
+    private ClusterState deleteDanglingIndex(ClusterState currentState, Index indexToDelete) {
+        final MetaData meta = currentState.metaData();
```

Comment:
do we need to double-check here that the index does not exist as proper index in the cluster state?